plugin-server/src/main/ingestion-queues/kafka-queue.ts

```diff
@@ -51,7 +59,64 @@ export class KafkaQueue implements Queue {
         }
     }
 
-    private async eachBatch({
+    private async eachMessageBuffer(
+        message: KafkaMessage,
+        resolveOffset: EachBatchPayload['resolveOffset']
+    ): Promise<void> {
+        const bufferEvent = JSON.parse(message.value!.toString())
+        await runInstrumentedFunction({
+            server: this.pluginsServer,
+            event: bufferEvent,
+            func: (_) => this.workerMethods.ingestBufferEvent(bufferEvent),
+            statsKey: `kafka_queue.ingest_buffer_event`,
+            timeoutMessage: 'After 30 seconds still running ingestBufferEvent',
+        })
+        resolveOffset(message.offset)
+    }
+
+    private async eachBatchBuffer({ batch, resolveOffset, commitOffsetsIfNecessary }: EachBatchPayload): Promise<void> {
+        if (batch.messages.length === 0) {
+            return
+        }
+        const batchStartTimer = new Date()
+
+        const promises = []
+        let consumerSleep = 0
+        for (const message of batch.messages) {
+            // kafka timestamps are unix timestamps in string format
+            const processAt = Number(message.timestamp) + this.pluginsServer.BUFFER_CONVERSION_SECONDS * 1000
+            const delayUntilTimeToProcess = processAt - new Date().getTime()
+
+            if (delayUntilTimeToProcess < 0) {
+                promises.push(this.eachMessageBuffer(message, resolveOffset))
+            } else {
+                consumerSleep = Math.max(consumerSleep, delayUntilTimeToProcess)
+            }
+        }
+
+        await Promise.all(promises)
+
+        // if consumerSleep > 0 it means we didn't process at least one message
+        if (consumerSleep > 0) {
+            // pause the consumer for this partition until we can process at least one message
+            await this.pause(batch.topic, batch.partition)
+            setTimeout(() => {
```

Comment:
I wonder if we should run this one before pause.

Reply:
fair yeah